src/hooks/progress/use-update-tutorial-progress.ts

```diff
@@ -0,0 +1,186 @@
+import {
+	useTutorialProgress,
+	useTutorialProgressMutations,
+	TutorialProgressMutationArgs,
+} from 'hooks/progress'
+// Types
+import {
+	TutorialIdCollectionId,
+	TutorialProgressStatus,
+} from 'lib/learn-client/types'
+
+type IntersectionRef = (node?: Element) => void
+
+interface UpdateProgressReturnInterface {
+	startRef: IntersectionRef
+	endRef: IntersectionRef
+}
+
+/**
+ * Given a tutorialId and collectionId,
+ * Return two refs in an object, { startRef, endRef },
+ * which can be placed on elements to track tutorial progress.
+ *
+ * On initial load, regardless of startRef or endRef visibility,
+ * if the tutorial does not yet have recorded progress,
+ * the tutorial will be marked as `visited` (0 percent).
+ *
+ * When the startRef element moves out of the viewport,
+ * the tutorial will be marked as `in_progress` (50 percent).
+ *
+ * When the endRef element moves into the viewport,
+ * the tutorial will be marked as `complete` (100 percent).
+ */
+export function useUpdateTutorialProgress({
+	tutorialId,
+	collectionId,
+}: TutorialIdCollectionId): UpdateProgressReturnInterface {
+	// We shouldn't try to update progress unless we're authenticated
+	const { isAuthenticated } = useAuthentication()
+	// We need to know if progress exists, to know whether to "create" or "update"
+	const { tutorialProgressStatus } = useTutorialProgress({
+		tutorialId,
+		collectionId,
+	})
+	// We'll use "create" or "update" progress using these `react-query` mutations
+	const tutorialProgressMutations = useTutorialProgressMutations()
+	// We use refs from `react-intersection-observer` to track progress.
+	const [startRef, startInView, startEntry] = useInView({
+		initialInView: true,
+	})
+	const [endRef, endInView, endEntry] = useInView()
+
+	/**
+	 * This effect runs often, but it seems necessary to ensure progress
+	 * gets tracked as things like the authenticated state changes.
+	 *
+	 * We bail early in many cases, and we avoid downgrading progress,
+	 * so this effect only results in meaningful mutations a small percent
+	 * of the time.
+	 */
+	useEffect(() => {
+		/**
+		 * On initial client-side navigation, this effect may be called
+		 * before startRefs and endRefs are updated. We can detect this
+		 * by comparing data-ref-id attributes we've added to the
+		 * elements startRef and endRef are attached to.
+		 */
+		const startRefId = startEntry?.target.getAttribute('data-ref-id')
+		const endRefId = endEntry?.target.getAttribute('data-ref-id')
+		const compareId = `${tutorialId}_${collectionId}`
+		const isStaticIdsMismatch =
+			startRefId !== compareId || endRefId !== compareId
+
+		console.log({
+			isStaticIdsMismatch,
+			startInView,
+			endInView,
+			startRefId,
+			endRefId,
+			compareId,
+		})
+		if (isStaticIdsMismatch) {
+			return
+		}
+
+		/**
+		 * If we're not authenticated, we bail early, there's no
+		 * point in trying to make a mutation.
+		 */
+		if (!isAuthenticated) {
+			return
+		}
```

Comment:
I think this should be very first in the block to really emphasize that nothing is happening if there is no authenticated user.

Reply:
After more reading, here's my full suggestion based on what you currently have coded (overly condensed just for the sake of pseudocode):

```typescript
useEffect(() => {
  // no computation needed to do the conditional check, var is already available
  if (!isAuthenticated) return
  
  // no computation needed to do the conditional check, var is already available
  if (typeof tutorialProgressStatus == 'undefined') return
  
  // requires computation to do the conditional, but less than the next block
  const isReadyToRunMutation = tutorialProgressMutations.isReady
  if (!isReadyToRunMutation) return
  
  // Requires the most computation to do conditional, so should be listed last in early bails
   const startRefId = startEntry?.target.getAttribute('data-ref-id')
  const endRefId = endEntry?.target.getAttribute('data-ref-id')
  const compareId = `${tutorialId}_${collectionId}`
  const isStaticIdsMismatch = startRefId !== compareId || endRefId !== compareId
  if (isStaticIdsMismatch) return
  
  // ... rest of the stuff
}, [])
```

---

What do you think about adding an additional early bail based on `tutorialProgressStatus`? As far as I can tell, we also don't need to do anything in this hook if `tutorialProgressStatus === 'complete'` since we can't have negative progress.

- If you do think it'd be good to add this, I'd place it second in the early bails list I've suggested, right after the `!isAuthenticated` one.

Reply:
✅ sweet, this makes a ton of sense to me, as does the additional early bail - have updated accordingly!